### DISCUSSION TITLE
Cannot Prisma Generate Optional Composite Foreign Key (for 1 to 0...1 Relation)

### DISCUSSION THREAD

**Discussion by ashih42** [prisma/prisma#3016] [Q&A] [answered]
Hello, I would like to model the relation that an `AdCreative` may have 0 or 1 `AdImage`, where an `AdImage` can be uniquely identified by the composite key `[hash, workspace_id]`.  My prisma schema is shown below.

```prisma
model AdImage {
  id                        String                @id  @default(cuid())

  hash                      String                // api hash

  ad_creatives              AdCreative[]

  workspace                 Workspace             @relation(fields: [workspace_id], references: [id])
  workspace_id              String

  @@unique([hash, workspace_id])
}

model AdCreative {
  id                        String                @id  @default(cuid())

  api_id                    String                // api id

  ad_image                  AdImage?              @relation(fields: [ad_image_hash, workspace_id], references: [hash, workspace_id])
  ad_image_hash             String?

  workspace                 Workspace             @relation(fields: [workspace_id], references: [id])
  workspace_id              String

  @@unique([api_id, workspace_id])
}
```

This is the error I got when running `prisma generate`:
```
Error: Schema parsing
error: Error validating: The relation field `ad_image` uses the scalar fields ad_image_hash, workspace_id. At least one of those fields is required. Hence the relation field must be required as well.
```

I think this error message means I am not allowed define `ad_image` as an optional field of type `AdImage?`.  Is there another way I can model this relation?

**Comment by ryands17** [comment]
Hey @ashih42 :wave: 
Currently that isn't possible by defining the model. Could you create the above relation in SQL and then run `npx prisma introspect` to see if it works?

If it doesn't even work then, I would suggest opening a request [here](https://github.com/prisma/prisma/issues/new/choose).

**Comment by ashih42** [comment, marked as answer]
Ugh, we decided to keep things simple by removing the schema definition `AdCreative.ad_image`, and manually write a resolver for `AdCreative.ad_image`.  This works for now.

**Reply by nikolasburk** [reply]
Good to hear you found a workaround! I marked this as an answer to your initial question, please let us know if you have further questions! 🙂

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
